pkg/reconciler/tenancy/initialization/apibinder_initializer_controller.go

```diff
@@ -0,0 +1,304 @@
+	createAPIBinding func(ctx context.Context, clusterName logicalcluster.Name, binding *apisv1alpha1.APIBinding) (*apisv1alpha1.APIBinding, error)
+	deleteAPIBinding func(ctx context.Context, clusterName logicalcluster.Name, bindingName string) error
+
+	createdBindingsLock sync.Mutex
+	createdBindings     map[logicalcluster.Name]map[tenancyv1alpha1.APIExportReference]struct{}
+
+	getAPIExport func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+
+	transitiveTypeResolver *admission.TransitiveTypeResolver
+
+	// commit creates a patch and submits it, if needed.
+	commit func(ctx context.Context, new, old *clusterWorkspaceResource) error
+}
+
+func (b *APIBinder) enqueueClusterWorkspace(obj interface{}, logger logr.Logger, logSuffix string) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logging.WithQueueKey(logger, key).V(2).Info(fmt.Sprintf("queueing ClusterWorkspace%s", logSuffix))
+	b.queue.Add(key)
+}
+
+func (b *APIBinder) enqueueAPIBinding(obj interface{}, logger logr.Logger) {
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("obj is supposed to be a metav1.Object, but is %T", obj))
+		return
+	}
+	clusterName := logicalcluster.From(metaObj)
+
+	clusterWorkspace, err := b.getClusterWorkspace(clusterName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// The workspace was deleted or is no longer initializing, so we can safely ignore this event.
+			return
+		}
+
+		parentClusterName, clusterWorkspaceName := clusters.SplitClusterAwareKey(clusterName.String())
+		logger.Error(err, "failed to get ClusterWorkspace from lister", "parentCluster", parentClusterName, "clusterWorkspace", clusterWorkspaceName)
+		return // nothing we can do here
+	}
+
+	b.enqueueClusterWorkspace(clusterWorkspace, logger, fmt.Sprintf(" because of APIBinding %s", metaObj.GetName()))
+}
+
+// enqueueClusterWorkspaceType enqueues all clusterworkspaces (which are only those that are initializing, because of
+// how the informer is supposed to be configured) whenever a clusterworkspacetype changes. If a clusterworkspacetype
+// had a typo in the default set of apibindings, there is a chance the requeuing here would pick up a fix.
+func (b *APIBinder) enqueueClusterWorkspaceType(obj interface{}, logger logr.Logger) {
+	cwt, ok := obj.(*tenancyv1alpha1.ClusterWorkspaceType)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("obj is supposed to be a ClusterWorkspaceType, but is %T", obj))
+		return
+	}
+
+	if len(cwt.Spec.DefaultAPIBindings) == 0 {
+		return
+	}
+
+	list, err := b.listClusterWorkspaces()
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("error listing clusterworkspaces: %w", err))
+	}
+
+	for _, ws := range list {
+		b.enqueueClusterWorkspace(ws, logger, fmt.Sprintf(" because of ClusterWorkspaceType %s|%s", logicalcluster.From(cwt), cwt.GetName()))
```

Comment:
similarly, attach the cwt to the logger, don't sprintf the fields of it

Reply:
will do